File: Sensor.Node/Messages/Client/AddModuleMessage.cs
using MongoDB.Bson;

namespace Sensor.Node.Messages.Client
{
    internal sealed class AddModuleMessage : ClientMessage
    {
        internal override State RequiredState
        {
            get { return State.Active; }
        }

        private int ModuleID { get; set; }
        //private int Resolution { get; set; }

        /// <summary>
        /// Initializes a new instance of the <see cref="AddModuleMessage"/> class.
        /// </summary>
        /// <param name="station"></param>
        internal AddModuleMessage(Station station) : base(station)
        {

        }

        internal override void Decode(BsonDocument document)
        {
            this.ModuleID = document["module_id"].AsInt32;
            //this.Resolution = document["res"].AsInt32;
        }

        internal override void Execute()
        {
            this.Station.ModuleManager.AddModule(this.ModuleID);
        }
    }
}

File: Sensor.Station/AddModuleMessage.cpp
#include "Module.hpp"
#include "AddModuleMessage.hpp"
#include "DigitalModule.hpp"

AddModuleMessage::AddModuleMessage(Module * module) : ClientMessage(module->station)
{
	this->id = 0x69;
	this->module = module;
}

void AddModuleMessage::Encode()
{
	ClientMessage::Encode();

	this->root["module_id"] = static_cast<int>(this->module->type);

	//if (!dynamic_cast<DigitalModule*>(this->module))
	//	this->root["res"] = this->module->resolution;
}


AddModuleMessage::~AddModuleMessage()
{
	this->module = nullptr;
}
